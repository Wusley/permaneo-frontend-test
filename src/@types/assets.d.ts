declare module '*.png';

declare module '*.svg' {
	import { ReactElement, SVGProps } from 'react'
	const ReactComponent: ( props: SVGProps<SVGElement> ) => ReactElement
	export { ReactComponent }
}

declare module '*.jpeg';

declare module '*.jpg';

declare module '*.gif';

declare module '*.webp';

declare module '*.ttf';

declare module '*.woff';

declare module '*.woff2';
