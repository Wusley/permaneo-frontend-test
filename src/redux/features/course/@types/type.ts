export interface Course {
	id: number
	thumbnail?: string
	title: string
	description: string
	price: number
	url: string
	startAt: string
	created_at: string
}

export interface CoursesSlice {
	course?: Course
	courses?: Course[]
	loading?: boolean
}
