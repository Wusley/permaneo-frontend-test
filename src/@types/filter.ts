export interface Filter {
	startDate: string
	endDate: string
}
