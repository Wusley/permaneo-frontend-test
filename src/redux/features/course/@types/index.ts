export * from './type'
export * from './response'
export * from './request'
