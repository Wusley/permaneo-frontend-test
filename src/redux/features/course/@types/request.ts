import { Course } from '.'

export interface CourseRequest {
	course?: Course
}
