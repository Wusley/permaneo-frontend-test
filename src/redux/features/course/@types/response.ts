import { Course } from '.'

export interface CourseResponse extends Response {
	data: Course
}

export interface CoursesResponse extends Response {
	data: Course[]
}