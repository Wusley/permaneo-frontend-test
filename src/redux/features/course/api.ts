import { CourseResponse, CoursesResponse } from './@types'
import api from '@/redux/services/api'

const courseAPI = api.injectEndpoints( {
  endpoints: ( builder ) => ( {
    listCourses: builder.query<CoursesResponse, { ids?: number[] }>( {
      query: ( { ids } ) => ( {
        url: '/courses',
        method: 'GET',
        params: ids ? { ids } : {}
      } )
    } ),
    course: builder.query<CourseResponse, { id: number }>( {
      query: ( { id } ) => ( {
        url: `/courses/${ id }`,
        method: 'GET'
      } )
    } )
  } )
} )

export const {
  useListCoursesQuery,
  useCourseQuery
} = courseAPI
