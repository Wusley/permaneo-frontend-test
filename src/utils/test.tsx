import { configureStore } from '@reduxjs/toolkit'
import { render } from '@testing-library/react'
import { renderHook } from '@testing-library/react'
import '@testing-library/jest-dom'
import Providers from '@/providers'
import { rootReducer, RootState } from '@/redux/store'

const createTestStore = ( preloadedState: Partial<RootState> = {} ) =>
  configureStore( {
    reducer: rootReducer,
    preloadedState,
    middleware: ( getDefaultMiddleware ) => getDefaultMiddleware()
  } )

const renderWithStore = (
  ui: React.ReactElement,
  preloadedState: Partial<RootState> = {},
  options?: unknown
) => {
  const store = createTestStore( preloadedState )

  return {
    ...render( ui, {
      wrapper: ( { children } ) => <Providers store={store}>{children}</Providers>,
      ...options
    } ),
    store
  }
}

const renderHookWithStore = <TProps, TResult>(
  hook: ( props: TProps ) => TResult,
  preloadedState: Partial<RootState> = {}
) => {
  const store = createTestStore( preloadedState )

  return {
    ...renderHook( hook, {
      wrapper: ( { children } ) => <Providers store={store}>{children}</Providers>
    } ),
    store
  }
}

export * from '@testing-library/react'
export { renderWithStore, renderHookWithStore }
