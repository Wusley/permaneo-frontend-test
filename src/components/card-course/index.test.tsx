import { screen, renderWithStore, fireEvent } from '@/utils/test'
import { CardCourse } from './'
import { Course } from '@/redux/features/course/@types'
import { useRouter } from 'next/navigation'

jest.mock( 'next/navigation', () => ( {
  useRouter: jest.fn()
} ) )

const mockCourse: Course = {
  id: 1,
  title: 'Curso de Next.js',
  price: 100,
  description: 'Aprenda Next.js do zero ao avanÃ§ado.',
  thumbnail: 'https://via.placeholder.com/150',
  created_at: '2021-10-01T00:00:00.000Z',
  url: 'https://www.youtube.com/embed/QhBnZ6NPOY0'
}

describe( 'CardCourse', () => {
  const mockPush = jest.fn()

  beforeEach( () => {
    ( useRouter as jest.Mock ).mockReturnValue( { push: mockPush } )
    mockPush.mockClear()
  } )

  it( 'renders the course card correctly', () => {
    const container = renderWithStore( <CardCourse course={mockCourse} isCourseActive={false} /> )

    expect( screen.getByRole( 'heading', { level: 3 } ) ).toHaveTextContent( mockCourse.title )
    expect( screen.getByText( mockCourse.description ) ).toBeInTheDocument()
    expect( screen.getByText( /inscreva-se agora/i ) ).toBeInTheDocument()
    expect( container ).toMatchSnapshot()
  } )

  it( 'renders active course state correctly', () => {
    const container = renderWithStore( <CardCourse course={mockCourse} isCourseActive={true} /> )

    expect( screen.getByText( /acessar curso/i ) ).toBeInTheDocument()
    expect( container ).toMatchSnapshot()
  } )

  it( 'navigates to active course page when clicked', () => {
    renderWithStore( <CardCourse course={mockCourse} isCourseActive={true} /> )
    fireEvent.click( screen.getByText( /acessar curso/i ) )
    expect( mockPush ).toHaveBeenCalledWith( `/curso/${mockCourse.id}` )
  } )

  it( 'navigates to the page details when clicked', () => {
    renderWithStore( <CardCourse course={mockCourse} isCourseActive={false} /> )
    fireEvent.click( screen.getByText( /inscreva-se agora/i ) )
    expect( mockPush ).toHaveBeenCalledWith( `/curso/detalhes/${mockCourse.id}` )
  } )
} )
