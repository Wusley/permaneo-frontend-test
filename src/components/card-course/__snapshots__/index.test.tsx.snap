// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardCourse renders active course state correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    class="chakra-ui-light"
  >
    <div>
      <div
        class="chakra-card css-1c6lmhu"
      >
        <header
          class="chakra-card__header css-1dad67h"
        >
          <div
            class="chakra-stack css-1uodvt1"
          >
            <span
              class="chakra-avatar css-tvcku"
            >
              <div
                aria-label="Curso de Next.js"
                class="chakra-avatar__initials css-1ebyn6"
                role="img"
              >
                CN
              </div>
            </span>
            <h3
              class="chakra-heading css-9f6g39"
            >
              Curso de Next.js
            </h3>
          </div>
        </header>
        <article
          class="chakra-card__body css-c8h2w1"
        >
          <p
            class="chakra-text css-tg4q1t"
          >
            Aprenda Next.js do zero ao avançado.
          </p>
        </article>
        <footer
          class="chakra-card__footer css-hoqr9a"
        >
          Acessar Curso
        </footer>
      </div>
      <span
        hidden=""
        id="__chakra_env"
      />
    </div>
    <div
      class="chakra-portal"
    >
      <div
        aria-label="Notifications-top"
        aria-live="polite"
        id="chakra-toast-manager-top"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-label="Notifications-top-left"
        aria-live="polite"
        id="chakra-toast-manager-top-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-top-right"
        aria-live="polite"
        id="chakra-toast-manager-top-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-bottom-left"
        aria-live="polite"
        id="chakra-toast-manager-bottom-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-bottom"
        aria-live="polite"
        id="chakra-toast-manager-bottom"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-label="Notifications-bottom-right"
        aria-live="polite"
        id="chakra-toast-manager-bottom-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="chakra-card css-1c6lmhu"
    >
      <header
        class="chakra-card__header css-1dad67h"
      >
        <div
          class="chakra-stack css-1uodvt1"
        >
          <span
            class="chakra-avatar css-tvcku"
          >
            <div
              aria-label="Curso de Next.js"
              class="chakra-avatar__initials css-1ebyn6"
              role="img"
            >
              CN
            </div>
          </span>
          <h3
            class="chakra-heading css-9f6g39"
          >
            Curso de Next.js
          </h3>
        </div>
      </header>
      <article
        class="chakra-card__body css-c8h2w1"
      >
        <p
          class="chakra-text css-tg4q1t"
        >
          Aprenda Next.js do zero ao avançado.
        </p>
      </article>
      <footer
        class="chakra-card__footer css-hoqr9a"
      >
        Acessar Curso
      </footer>
    </div>
    <span
      hidden=""
      id="__chakra_env"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;

exports[`CardCourse renders the course card correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    class="chakra-ui-light"
  >
    <div>
      <div
        class="chakra-card css-1fxlf1t"
      >
        <header
          class="chakra-card__header css-z813yp"
        >
          <div
            class="chakra-stack css-1uodvt1"
          >
            <span
              class="chakra-avatar css-tvcku"
            >
              <div
                aria-label="Curso de Next.js"
                class="chakra-avatar__initials css-1ebyn6"
                role="img"
              >
                CN
              </div>
            </span>
            <h3
              class="chakra-heading css-9f6g39"
            >
              Curso de Next.js
            </h3>
          </div>
        </header>
        <article
          class="chakra-card__body css-1ttha67"
        >
          <p
            class="chakra-text css-tg4q1t"
          >
            Aprenda Next.js do zero ao avançado.
          </p>
        </article>
        <footer
          class="chakra-card__footer css-rfvz8g"
        >
          Inscreva-se agora
        </footer>
      </div>
      <span
        hidden=""
        id="__chakra_env"
      />
    </div>
    <div
      class="chakra-portal"
    >
      <div
        aria-label="Notifications-top"
        aria-live="polite"
        id="chakra-toast-manager-top"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-label="Notifications-top-left"
        aria-live="polite"
        id="chakra-toast-manager-top-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-top-right"
        aria-live="polite"
        id="chakra-toast-manager-top-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-bottom-left"
        aria-live="polite"
        id="chakra-toast-manager-bottom-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-bottom"
        aria-live="polite"
        id="chakra-toast-manager-bottom"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-label="Notifications-bottom-right"
        aria-live="polite"
        id="chakra-toast-manager-bottom-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="chakra-card css-1fxlf1t"
    >
      <header
        class="chakra-card__header css-z813yp"
      >
        <div
          class="chakra-stack css-1uodvt1"
        >
          <span
            class="chakra-avatar css-tvcku"
          >
            <div
              aria-label="Curso de Next.js"
              class="chakra-avatar__initials css-1ebyn6"
              role="img"
            >
              CN
            </div>
          </span>
          <h3
            class="chakra-heading css-9f6g39"
          >
            Curso de Next.js
          </h3>
        </div>
      </header>
      <article
        class="chakra-card__body css-1ttha67"
      >
        <p
          class="chakra-text css-tg4q1t"
        >
          Aprenda Next.js do zero ao avançado.
        </p>
      </article>
      <footer
        class="chakra-card__footer css-rfvz8g"
      >
        Inscreva-se agora
      </footer>
    </div>
    <span
      hidden=""
      id="__chakra_env"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
